warp/aggregator/aggregator.go

```diff
@@ -36,16 +71,150 @@ func (a *Aggregator) AggregateSignatures(ctx context.Context, unsignedMessage *a
+	quorumNum uint64,
+	totalWeight uint64,
+) (*AggregateSignatureResult, error) {
+	type signatureFetchResult struct {
+		sig    *bls.Signature
+		index  int
+		weight uint64
+	}
+
+	// Create a child context to cancel signature fetching if we reach signature threshold.
+	signatureFetchCtx, signatureFetchCancel := context.WithCancel(ctx)
+	defer signatureFetchCancel()
+
+	// Fetch signatures from validators concurrently.
+	signatureFetchResultChan := make(chan *signatureFetchResult, len(validators))
+	for i, validator := range validators {
+		var (
+			i         = i
+			validator = validator
+			// TODO: update from a single nodeID to the original slice and use extra nodeIDs as backup.
+			nodeID = validator.NodeIDs[0]
+		)
+		go func() {
+			log.Debug("Fetching warp signature",
+				"nodeID", nodeID,
+				"index", i,
+			)
+
+			signature, err := client.GetSignature(signatureFetchCtx, nodeID, unsignedMessage)
+			if err != nil {
+				log.Debug("Failed to fetch warp signature",
+					"nodeID", nodeID,
+					"index", i,
+					"err", err,
+				)
+				signatureFetchResultChan <- nil
+				return
+			}
+
+			sigHex := hexutil.Bytes(bls.SignatureToBytes(signature))
+			log.Debug("Retrieved warp signature",
+				"nodeID", nodeID,
+				"index", i,
+				"signature", sigHex,
+			)
+
+			if !bls.Verify(validator.PublicKey, signature, unsignedMessage.Bytes()) {
+				log.Debug("Failed to verify warp signature",
+					"nodeID", nodeID,
+					"index", i,
+					"signature", sigHex,
+					"msgID", unsignedMessage.ID(),
+				)
+				signatureFetchResultChan <- nil
+				return
+			}
+
+			signatureFetchResultChan <- &signatureFetchResult{
+				sig:    signature,
+				index:  i,
+				weight: validator.Weight,
+			}
+		}()
+	}
+
+	var (
+		signatures                = make([]*bls.Signature, 0, len(validators))
+		signersBitset             = set.NewBits()
+		signaturesWeight          = uint64(0)
+		signaturesPassedThreshold = false
+	)
+
+	for i := 0; i < len(validators); i++ {
+		signatureFetchResult := <-signatureFetchResultChan
+		if signatureFetchResult == nil {
+			continue
+		}
+
+		signatures = append(signatures, signatureFetchResult.sig)
+		signersBitset.Add(signatureFetchResult.index)
+		signaturesWeight += signatureFetchResult.weight
+		log.Debug("Updated weight",
+			"totalWeight", signaturesWeight,
+			"addedWeight", signatureFetchResult.weight,
+		)
+
+		// If the signature weight meets the requested threshold, cancel signature fetching
+		if err := avalancheWarp.VerifyWeight(signaturesWeight, totalWeight, quorumNum, params.WarpQuorumDenominator); err == nil {
+			log.Debug("Verify weight passed, exiting aggregation early",
+				"quorumNum", quorumNum,
+				"totalWeight", totalWeight,
+				"signatureWeight", signaturesWeight,
+			)
+			signatureFetchCancel()
+			signaturesPassedThreshold = true
+			break
+		}
+	}
+
+	// If I failed to fetch sufficient signature stake, return an error
+	if !signaturesPassedThreshold {
+		return nil, errInsufficientWeight
```

Comment:
should we use avalancheWarp.ErrInsufficientWeight here?

Reply:
good catch